Deployment tests for networks don't pass

I was trying to run deployment tests for #702 #683 but those were failing. So I decided to run the tests with what we currently have in the master and the tests failed for all the switches. The deployment tests only passed with mock drivers. 

I have put the logs in the comments here; the tests are getting stuck at `create_networks()`

I can share my `site-layout.json` and `testsuite.cfg` files if you feel like testing. 

When was the last time we ran the network deployment tests with real hardware?

Here are the results for the dell

```
(.venv)naved001:~/mock-deploy-test/hil$ py.test tests/deployment/*networks.py
========================================= test session starts =========================================
platform linux2 -- Python 2.7.5, pytest-2.9.2, py-1.4.32, pluggy-0.3.1
rootdir: /home/naved001/mock-deploy-test/hil, inifile: setup.cfg
plugins: xdist-1.15.0, cov-1.8.1
collected 2 items 

tests/deployment/native_networks.py F
tests/deployment/vlan_networks.py F

============================================== FAILURES ===============================================
______________________________ TestNativeNetwork.test_isolated_networks _______________________________
self = <native_networks.TestNativeNetwork instance at 0x27ed248>

    def test_isolated_networks(self):
    
        def create_networks():
            nodes = self.collect_nodes()
    
            # Create two networks
            network_create_simple('net-0', 'anvil-nextgen')
            network_create_simple('net-1', 'anvil-nextgen')
    
            ports = self.get_all_ports(nodes)
    
            # Assert that n0 and n1 are not on any network
            port_networks = self.get_port_networks(ports)
    
            assert self.get_network(nodes[0].nics[0].port, port_networks) == \
                set()
            assert self.get_network(nodes[1].nics[0].port, port_networks) == \
                set()
    
            # Connect n0 and n1 to net-0 and net-1 respectively
            api.node_connect_network(nodes[0].label,
                                     nodes[0].nics[0].label,
                                     'net-0')
            api.node_connect_network(nodes[1].label,
                                     nodes[1].nics[0].label,
                                     'net-1')
            deferred.apply_networking()
    
            # Assert that n0 and n1 are on isolated networks
            port_networks = self.get_port_networks(ports)
            assert self.get_network(nodes[0].nics[0].port, port_networks) == \
                set([nodes[0].nics[0].port])
            assert self.get_network(nodes[1].nics[0].port, port_networks) == \
                set([nodes[1].nics[0].port])
    
            # Add n2 and n3 to the same networks as n0 and n1 respectively
            api.node_connect_network(nodes[2].label,
                                     nodes[2].nics[0].label,
                                     'net-0')
            api.node_connect_network(nodes[3].label,
                                     nodes[3].nics[0].label,
                                     'net-1')
            deferred.apply_networking()
    
            # Assert that n2 and n3 have been added to n0 and n1's networks
            # respectively
            port_networks = self.get_port_networks(ports)
            assert self.get_network(nodes[0].nics[0].port, port_networks) == \
                set([nodes[0].nics[0].port, nodes[2].nics[0].port])
            assert self.get_network(nodes[1].nics[0].port, port_networks) == \
                set([nodes[1].nics[0].port, nodes[3].nics[0].port])
    
        def delete_networks():
            # Query the DB for nodes on this project
            project = api._must_find(model.Project, 'anvil-nextgen')
            nodes = project.nodes
            ports = self.get_all_ports(nodes)
    
            # Remove all nodes from their networks
            for node in nodes:
                attachment = model.NetworkAttachment.query \
                    .filter_by(nic=node.nics[0]).one()
                api.node_detach_network(node.label,
                                        node.nics[0].label,
                                        attachment.network.label)
            deferred.apply_networking()
    
            # Assert that none of the nodes are on any network
            port_networks = self.get_port_networks(ports)
            for node in nodes:
                assert self.get_network(node.nics[0].port, port_networks) == \
                    set()
    
            # Delete the networks
            api.network_delete('net-0')
            api.network_delete('net-1')
    
        # Create a project
        api.project_create('anvil-nextgen')
    
>       create_networks()

tests/deployment/native_networks.py:135: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
tests/deployment/native_networks.py:67: in create_networks
    port_networks = self.get_port_networks(ports)
haas/test_common.py:228: in get_port_networks
    switch_port_networks = session.get_port_networks(ports)
haas/ext/switches/dell.py:122: in get_port_networks
    port_configs = self._port_configs(ports)
haas/ext/switches/dell.py:155: in _port_configs
    result[port] = self._int_config(port.label)
haas/ext/switches/dell.py:180: in _int_config
    index = self.console.expect(alternatives)
.venv/lib/python2.7/site-packages/pexpect/__init__.py:1451: in expect
    timeout, searchwindowsize)
.venv/lib/python2.7/site-packages/pexpect/__init__.py:1466: in expect_list
    timeout, searchwindowsize)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <pexpect.spawn object at 0x2db8110>, searcher = <pexpect.searcher_re object at 0x2dbc410>
timeout = 29.994247913360596, searchwindowsize = None

    def expect_loop(self, searcher, timeout=-1, searchwindowsize=-1):
    
        '''This is the common loop used inside expect. The 'searcher' should be
            an instance of searcher_re or searcher_string, which describes how and
            what to search for in the input.
    
            See expect() for other arguments, return value and exceptions. '''
    
        self.searcher = searcher
    
        if timeout == -1:
            timeout = self.timeout
        if timeout is not None:
            end_time = time.time() + timeout
        if searchwindowsize == -1:
            searchwindowsize = self.searchwindowsize
    
        try:
            incoming = self.buffer
            freshlen = len(incoming)
            while True:
                # Keep reading until exception or return.
                index = searcher.search(incoming, freshlen, searchwindowsize)
                if index >= 0:
                    self.buffer = incoming[searcher.end:]
                    self.before = incoming[: searcher.start]
                    self.after = incoming[searcher.start: searcher.end]
                    self.match = searcher.match
                    self.match_index = index
                    return self.match_index
                # No match at this point
                if (timeout is not None) and (timeout < 0):
                    raise TIMEOUT('Timeout exceeded in expect_any().')
                # Still have time left, so read more data
                c = self.read_nonblocking(self.maxread, timeout)
                freshlen = len(c)
                time.sleep(0.0001)
                incoming = incoming + c
                if timeout is not None:
                    timeout = end_time - time.time()
        except EOF:
            err = sys.exc_info()[1]
            self.buffer = self.string_type()
            self.before = incoming
            self.after = EOF
            index = searcher.eof_index
            if index >= 0:
                self.match = EOF
                self.match_index = index
                return self.match_index
            else:
                self.match = None
                self.match_index = None
                raise EOF(str(err) + '\n' + str(self))
        except TIMEOUT:
            err = sys.exc_info()[1]
            self.buffer = incoming
            self.before = incoming
            self.after = TIMEOUT
            index = searcher.timeout_index
            if index >= 0:
                self.match = TIMEOUT
                self.match_index = index
                return self.match_index
            else:
                self.match = None
                self.match_index = None
>               raise TIMEOUT(str(err) + '\n' + str(self))
E               TIMEOUT: Timeout exceeded.
E               <pexpect.spawn object at 0x2db8110>
E               version: 3.3
E               command: /usr/bin/telnet
E               args: [u'/usr/bin/telnet', u'192.168.3.247']
E               searcher: <pexpect.searcher_re object at 0x2dbc410>
E               buffer (last 100 chars): '\x1b[0mMore: <space>,  Quit: q or CTRL+Z, One line: <return> '
E               before (last 100 chars): '\x1b[0mMore: <space>,  Quit: q or CTRL+Z, One line: <return> '
E               after: <class 'pexpect.TIMEOUT'>
E               match: None
E               match_index: None
E               exitstatus: None
E               flag_eof: False
E               pid: 12075
E               child_fd: 13
E               closed: False
E               timeout: 30
E               delimiter: <class 'pexpect.EOF'>
E               logfile: None
E               logfile_read: None
E               logfile_send: None
E               maxread: 2000
E               ignorecase: False
E               searchwindowsize: None
E               delaybeforesend: 0.05
E               delayafterclose: 0.1
E               delayafterterminate: 0.1

.venv/lib/python2.7/site-packages/pexpect/__init__.py:1568: TIMEOUT
---------------------------------------- Captured stderr setup ----------------------------------------
INFO  [alembic.runtime.migration] Context impl SQLiteImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.runtime.migration] Running stamp_revision  -> 57f4c30b0ad4
INFO  [alembic.runtime.migration] Context impl SQLiteImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.runtime.migration] Running stamp_revision  -> df8d9f423f2b
INFO  [alembic.runtime.migration] Context impl SQLiteImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.runtime.migration] Running stamp_revision  -> 099b939261c1
_______________________________ TestNetworkVlan.test_isolated_networks ________________________________
self = <vlan_networks.TestNetworkVlan instance at 0x27e72d8>

    def test_isolated_networks(self):
    
        def get_legal_channels(network):
            response_body = api.show_network(network)
            response_body = json.loads(response_body)
            return response_body['channels']
    
        def create_networks():
            nodes = self.collect_nodes()
    
            # Create two networks
            network_create_simple('net-0', 'anvil-nextgen')
            network_create_simple('net-1', 'anvil-nextgen')
    
            ports = self.get_all_ports(nodes)
    
            # Assert that n0 and n1 are not on any network
            port_networks = self.get_port_networks(ports)
    
            assert self.get_network(nodes[0].nics[0].port, port_networks) == \
                set()
            assert self.get_network(nodes[1].nics[0].port, port_networks) == \
                set()
    
            # Get the channel ids for the tagged versions of the networks:
            net_tag = {}
            net_tag[0] = get_legal_channels('net-0')[1]
            net_tag[1] = get_legal_channels('net-1')[1]
    
            # Connect node 0 to net-0 (native mode)
            api.node_connect_network(nodes[0].label,
                                     nodes[0].nics[0].label,
                                     'net-0')
            # Connect node 1 to net-1 (tagged mode)
            api.node_connect_network(nodes[1].label,
                                     nodes[1].nics[0].label,
                                     'net-1',
                                     channel=net_tag[1])
            deferred.apply_networking()
    
            # Assert that n0 and n1 are on isolated networks
            port_networks = self.get_port_networks(ports)
            assert self.get_network(nodes[0].nics[0].port, port_networks) == \
                set([nodes[0].nics[0].port])
            assert self.get_network(nodes[1].nics[0].port, port_networks) == \
                set([nodes[1].nics[0].port])
    
            # Add n2 and n3 to the same networks as n0 and n1 respectively, but
            # with different channels (native vs. tagged)
            api.node_connect_network(nodes[2].label,
                                     nodes[2].nics[0].label,
                                     'net-0',
                                     channel=net_tag[0])
            api.node_connect_network(nodes[3].label,
                                     nodes[3].nics[0].label,
                                     'net-1')
            deferred.apply_networking()
    
            # Assert that n2 and n3 have been added to n0 and n1's networks
            # respectively
            port_networks = self.get_port_networks(ports)
            assert self.get_network(nodes[0].nics[0].port, port_networks) == \
                set([nodes[0].nics[0].port, nodes[2].nics[0].port])
            assert self.get_network(nodes[1].nics[0].port, port_networks) == \
                set([nodes[1].nics[0].port, nodes[3].nics[0].port])
    
            # Verify that we can put nodes on more than one network, with
            # different channels:
            api.node_connect_network(nodes[2].label,
                                     nodes[2].nics[0].label,
                                     'net-1')
            deferred.apply_networking()
            port_networks = self.get_port_networks(ports)
            assert self.get_network(nodes[1].nics[0].port, port_networks) == \
                set([nodes[1].nics[0].port,
                     nodes[2].nics[0].port,
                     nodes[3].nics[0].port])
    
        def delete_networks():
            # Query the DB for nodes on this project
            project = api._must_find(model.Project, 'anvil-nextgen')
            nodes = project.nodes
            ports = self.get_all_ports(nodes)
    
            # Remove all nodes from their networks. We first build up a list of
            # the arguments to the API calls, which has no direct references to
            # database objects, and then make the API calls and invoke
            # deferred.apply_networking after. This is important --
            # The API calls and apply_networking normally run in their own
            # transaction. We get away with not doing this in the tests because
            # we serialize everything ourselves, so there's no risk of
            # interference. If we were to hang on to references to database
            # objects across such calls however, things could get harry.
            all_attachments = []
            for node in nodes:
                attachments = model.NetworkAttachment.query \
                    .filter_by(nic=node.nics[0]).all()
                for attachment in attachments:
                    all_attachments.append((node.label,
                                            node.nics[0].label,
                                            attachment.network.label))
            for attachment in all_attachments:
                api.node_detach_network(*attachment)
                deferred.apply_networking()
    
            # Assert that none of the nodes are on any network
            port_networks = self.get_port_networks(ports)
            for node in nodes:
                assert self.get_network(node.nics[0].port, port_networks) == \
                    set()
    
            # Delete the networks
            api.network_delete('net-0')
            api.network_delete('net-1')
    
        # Create a project
        api.project_create('anvil-nextgen')
    
>       create_networks()

tests/deployment/vlan_networks.py:171: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
tests/deployment/vlan_networks.py:70: in create_networks
    port_networks = self.get_port_networks(ports)
haas/test_common.py:228: in get_port_networks
    switch_port_networks = session.get_port_networks(ports)
haas/ext/switches/dell.py:122: in get_port_networks
    port_configs = self._port_configs(ports)
haas/ext/switches/dell.py:155: in _port_configs
    result[port] = self._int_config(port.label)
haas/ext/switches/dell.py:180: in _int_config
    index = self.console.expect(alternatives)
.venv/lib/python2.7/site-packages/pexpect/__init__.py:1451: in expect
    timeout, searchwindowsize)
.venv/lib/python2.7/site-packages/pexpect/__init__.py:1466: in expect_list
    timeout, searchwindowsize)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <pexpect.spawn object at 0x2e86dd0>, searcher = <pexpect.searcher_re object at 0x2eaa310>
timeout = 29.99424386024475, searchwindowsize = None

    def expect_loop(self, searcher, timeout=-1, searchwindowsize=-1):
    
        '''This is the common loop used inside expect. The 'searcher' should be
            an instance of searcher_re or searcher_string, which describes how and
            what to search for in the input.
    
            See expect() for other arguments, return value and exceptions. '''
    
        self.searcher = searcher
    
        if timeout == -1:
            timeout = self.timeout
        if timeout is not None:
            end_time = time.time() + timeout
        if searchwindowsize == -1:
            searchwindowsize = self.searchwindowsize
    
        try:
            incoming = self.buffer
            freshlen = len(incoming)
            while True:
                # Keep reading until exception or return.
                index = searcher.search(incoming, freshlen, searchwindowsize)
                if index >= 0:
                    self.buffer = incoming[searcher.end:]
                    self.before = incoming[: searcher.start]
                    self.after = incoming[searcher.start: searcher.end]
                    self.match = searcher.match
                    self.match_index = index
                    return self.match_index
                # No match at this point
                if (timeout is not None) and (timeout < 0):
                    raise TIMEOUT('Timeout exceeded in expect_any().')
                # Still have time left, so read more data
                c = self.read_nonblocking(self.maxread, timeout)
                freshlen = len(c)
                time.sleep(0.0001)
                incoming = incoming + c
                if timeout is not None:
                    timeout = end_time - time.time()
        except EOF:
            err = sys.exc_info()[1]
            self.buffer = self.string_type()
            self.before = incoming
            self.after = EOF
            index = searcher.eof_index
            if index >= 0:
                self.match = EOF
                self.match_index = index
                return self.match_index
            else:
                self.match = None
                self.match_index = None
                raise EOF(str(err) + '\n' + str(self))
        except TIMEOUT:
            err = sys.exc_info()[1]
            self.buffer = incoming
            self.before = incoming
            self.after = TIMEOUT
            index = searcher.timeout_index
            if index >= 0:
                self.match = TIMEOUT
                self.match_index = index
                return self.match_index
            else:
                self.match = None
                self.match_index = None
>               raise TIMEOUT(str(err) + '\n' + str(self))
E               TIMEOUT: Timeout exceeded.
E               <pexpect.spawn object at 0x2e86dd0>
E               version: 3.3
E               command: /usr/bin/telnet
E               args: [u'/usr/bin/telnet', u'192.168.3.247']
E               searcher: <pexpect.searcher_re object at 0x2eaa310>
E               buffer (last 100 chars): '\x1b[0mMore: <space>,  Quit: q or CTRL+Z, One line: <return> '
E               before (last 100 chars): '\x1b[0mMore: <space>,  Quit: q or CTRL+Z, One line: <return> '
E               after: <class 'pexpect.TIMEOUT'>
E               match: None
E               match_index: None
E               exitstatus: None
E               flag_eof: False
E               pid: 12085
E               child_fd: 13
E               closed: False
E               timeout: 30
E               delimiter: <class 'pexpect.EOF'>
E               logfile: None
E               logfile_read: None
E               logfile_send: None
E               maxread: 2000
E               ignorecase: False
E               searchwindowsize: None
E               delaybeforesend: 0.05
E               delayafterclose: 0.1
E               delayafterterminate: 0.1

.venv/lib/python2.7/site-packages/pexpect/__init__.py:1568: TIMEOUT
---------------------------------------- Captured stderr setup ----------------------------------------
INFO  [alembic.runtime.migration] Context impl SQLiteImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.runtime.migration] Running stamp_revision  -> 57f4c30b0ad4
INFO  [alembic.runtime.migration] Context impl SQLiteImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.runtime.migration] Running stamp_revision  -> df8d9f423f2b
INFO  [alembic.runtime.migration] Context impl SQLiteImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.runtime.migration] Running stamp_revision  -> 099b939261c1
====================================== 2 failed in 89.65 seconds ======================================
```

Here are the results for the brocade:

```
(.venv)naved001:~/mock-deploy-test/hil$ py.test tests/deployment/*networks.py
========================================= test session starts =========================================
platform linux2 -- Python 2.7.5, pytest-2.9.2, py-1.4.32, pluggy-0.3.1
rootdir: /home/naved001/mock-deploy-test/hil, inifile: setup.cfg
plugins: xdist-1.15.0, cov-1.8.1
collected 2 items 

tests/deployment/native_networks.py F
tests/deployment/vlan_networks.py F

============================================== FAILURES ===============================================
______________________________ TestNativeNetwork.test_isolated_networks _______________________________
self = <native_networks.TestNativeNetwork instance at 0x3024248>

    def test_isolated_networks(self):
    
        def create_networks():
            nodes = self.collect_nodes()
    
            # Create two networks
            network_create_simple('net-0', 'anvil-nextgen')
            network_create_simple('net-1', 'anvil-nextgen')
    
            ports = self.get_all_ports(nodes)
    
            # Assert that n0 and n1 are not on any network
            port_networks = self.get_port_networks(ports)
    
            assert self.get_network(nodes[0].nics[0].port, port_networks) == \
                set()
            assert self.get_network(nodes[1].nics[0].port, port_networks) == \
                set()
    
            # Connect n0 and n1 to net-0 and net-1 respectively
            api.node_connect_network(nodes[0].label,
                                     nodes[0].nics[0].label,
                                     'net-0')
            api.node_connect_network(nodes[1].label,
                                     nodes[1].nics[0].label,
                                     'net-1')
            deferred.apply_networking()
    
            # Assert that n0 and n1 are on isolated networks
            port_networks = self.get_port_networks(ports)
            assert self.get_network(nodes[0].nics[0].port, port_networks) == \
                set([nodes[0].nics[0].port])
            assert self.get_network(nodes[1].nics[0].port, port_networks) == \
                set([nodes[1].nics[0].port])
    
            # Add n2 and n3 to the same networks as n0 and n1 respectively
            api.node_connect_network(nodes[2].label,
                                     nodes[2].nics[0].label,
                                     'net-0')
            api.node_connect_network(nodes[3].label,
                                     nodes[3].nics[0].label,
                                     'net-1')
            deferred.apply_networking()
    
            # Assert that n2 and n3 have been added to n0 and n1's networks
            # respectively
            port_networks = self.get_port_networks(ports)
            assert self.get_network(nodes[0].nics[0].port, port_networks) == \
                set([nodes[0].nics[0].port, nodes[2].nics[0].port])
            assert self.get_network(nodes[1].nics[0].port, port_networks) == \
                set([nodes[1].nics[0].port, nodes[3].nics[0].port])
    
        def delete_networks():
            # Query the DB for nodes on this project
            project = api._must_find(model.Project, 'anvil-nextgen')
            nodes = project.nodes
            ports = self.get_all_ports(nodes)
    
            # Remove all nodes from their networks
            for node in nodes:
                attachment = model.NetworkAttachment.query \
                    .filter_by(nic=node.nics[0]).one()
                api.node_detach_network(node.label,
                                        node.nics[0].label,
                                        attachment.network.label)
            deferred.apply_networking()
    
            # Assert that none of the nodes are on any network
            port_networks = self.get_port_networks(ports)
            for node in nodes:
                assert self.get_network(node.nics[0].port, port_networks) == \
                    set()
    
            # Delete the networks
            api.network_delete('net-0')
            api.network_delete('net-1')
    
        # Create a project
        api.project_create('anvil-nextgen')
    
>       create_networks()

tests/deployment/native_networks.py:135: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
tests/deployment/native_networks.py:67: in create_networks
    port_networks = self.get_port_networks(ports)
haas/test_common.py:228: in get_port_networks
    switch_port_networks = session.get_port_networks(ports)
haas/ext/switches/brocade.py:107: in get_port_networks
    response[port] = filter(None, [self._get_native_vlan(port)]) \
haas/ext/switches/brocade.py:179: in _get_native_vlan
    root = etree.fromstring(response.text)
src/lxml/lxml.etree.pyx:3213: in lxml.etree.fromstring (src/lxml/lxml.etree.c:79010)
    ???
src/lxml/parser.pxi:1848: in lxml.etree._parseMemoryDocument (src/lxml/lxml.etree.c:118341)
    ???
src/lxml/parser.pxi:1729: in lxml.etree._parseDoc (src/lxml/lxml.etree.c:116899)
    ???
src/lxml/parser.pxi:1063: in lxml.etree._BaseParser._parseUnicodeDoc (src/lxml/lxml.etree.c:110886)
    ???
src/lxml/parser.pxi:595: in lxml.etree._ParserContext._handleParseResultDoc (src/lxml/lxml.etree.c:105109)
    ???
src/lxml/parser.pxi:706: in lxml.etree._handleParseResult (src/lxml/lxml.etree.c:106817)
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   XMLSyntaxError: None (line 0)

src/lxml/parser.pxi:646: XMLSyntaxError
---------------------------------------- Captured stderr setup ----------------------------------------
INFO  [alembic.runtime.migration] Context impl SQLiteImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.runtime.migration] Running stamp_revision  -> 57f4c30b0ad4
INFO  [alembic.runtime.migration] Context impl SQLiteImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.runtime.migration] Running stamp_revision  -> df8d9f423f2b
INFO  [alembic.runtime.migration] Context impl SQLiteImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.runtime.migration] Running stamp_revision  -> 5a6db7a7222d
_______________________________ TestNetworkVlan.test_isolated_networks ________________________________
self = <vlan_networks.TestNetworkVlan instance at 0x301e2d8>

    def test_isolated_networks(self):
    
        def get_legal_channels(network):
            response_body = api.show_network(network)
            response_body = json.loads(response_body)
            return response_body['channels']
    
        def create_networks():
            nodes = self.collect_nodes()
    
            # Create two networks
            network_create_simple('net-0', 'anvil-nextgen')
            network_create_simple('net-1', 'anvil-nextgen')
    
            ports = self.get_all_ports(nodes)
    
            # Assert that n0 and n1 are not on any network
            port_networks = self.get_port_networks(ports)
    
            assert self.get_network(nodes[0].nics[0].port, port_networks) == \
                set()
            assert self.get_network(nodes[1].nics[0].port, port_networks) == \
                set()
    
            # Get the channel ids for the tagged versions of the networks:
            net_tag = {}
            net_tag[0] = get_legal_channels('net-0')[1]
            net_tag[1] = get_legal_channels('net-1')[1]
    
            # Connect node 0 to net-0 (native mode)
            api.node_connect_network(nodes[0].label,
                                     nodes[0].nics[0].label,
                                     'net-0')
            # Connect node 1 to net-1 (tagged mode)
            api.node_connect_network(nodes[1].label,
                                     nodes[1].nics[0].label,
                                     'net-1',
                                     channel=net_tag[1])
            deferred.apply_networking()
    
            # Assert that n0 and n1 are on isolated networks
            port_networks = self.get_port_networks(ports)
            assert self.get_network(nodes[0].nics[0].port, port_networks) == \
                set([nodes[0].nics[0].port])
            assert self.get_network(nodes[1].nics[0].port, port_networks) == \
                set([nodes[1].nics[0].port])
    
            # Add n2 and n3 to the same networks as n0 and n1 respectively, but
            # with different channels (native vs. tagged)
            api.node_connect_network(nodes[2].label,
                                     nodes[2].nics[0].label,
                                     'net-0',
                                     channel=net_tag[0])
            api.node_connect_network(nodes[3].label,
                                     nodes[3].nics[0].label,
                                     'net-1')
            deferred.apply_networking()
    
            # Assert that n2 and n3 have been added to n0 and n1's networks
            # respectively
            port_networks = self.get_port_networks(ports)
            assert self.get_network(nodes[0].nics[0].port, port_networks) == \
                set([nodes[0].nics[0].port, nodes[2].nics[0].port])
            assert self.get_network(nodes[1].nics[0].port, port_networks) == \
                set([nodes[1].nics[0].port, nodes[3].nics[0].port])
    
            # Verify that we can put nodes on more than one network, with
            # different channels:
            api.node_connect_network(nodes[2].label,
                                     nodes[2].nics[0].label,
                                     'net-1')
            deferred.apply_networking()
            port_networks = self.get_port_networks(ports)
            assert self.get_network(nodes[1].nics[0].port, port_networks) == \
                set([nodes[1].nics[0].port,
                     nodes[2].nics[0].port,
                     nodes[3].nics[0].port])
    
        def delete_networks():
            # Query the DB for nodes on this project
            project = api._must_find(model.Project, 'anvil-nextgen')
            nodes = project.nodes
            ports = self.get_all_ports(nodes)
    
            # Remove all nodes from their networks. We first build up a list of
            # the arguments to the API calls, which has no direct references to
            # database objects, and then make the API calls and invoke
            # deferred.apply_networking after. This is important --
            # The API calls and apply_networking normally run in their own
            # transaction. We get away with not doing this in the tests because
            # we serialize everything ourselves, so there's no risk of
            # interference. If we were to hang on to references to database
            # objects across such calls however, things could get harry.
            all_attachments = []
            for node in nodes:
                attachments = model.NetworkAttachment.query \
                    .filter_by(nic=node.nics[0]).all()
                for attachment in attachments:
                    all_attachments.append((node.label,
                                            node.nics[0].label,
                                            attachment.network.label))
            for attachment in all_attachments:
                api.node_detach_network(*attachment)
                deferred.apply_networking()
    
            # Assert that none of the nodes are on any network
            port_networks = self.get_port_networks(ports)
            for node in nodes:
                assert self.get_network(node.nics[0].port, port_networks) == \
                    set()
    
            # Delete the networks
            api.network_delete('net-0')
            api.network_delete('net-1')
    
        # Create a project
        api.project_create('anvil-nextgen')
    
>       create_networks()

tests/deployment/vlan_networks.py:171: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
tests/deployment/vlan_networks.py:70: in create_networks
    port_networks = self.get_port_networks(ports)
haas/test_common.py:228: in get_port_networks
    switch_port_networks = session.get_port_networks(ports)
haas/ext/switches/brocade.py:107: in get_port_networks
    response[port] = filter(None, [self._get_native_vlan(port)]) \
haas/ext/switches/brocade.py:179: in _get_native_vlan
    root = etree.fromstring(response.text)
src/lxml/lxml.etree.pyx:3213: in lxml.etree.fromstring (src/lxml/lxml.etree.c:79010)
    ???
src/lxml/parser.pxi:1848: in lxml.etree._parseMemoryDocument (src/lxml/lxml.etree.c:118341)
    ???
src/lxml/parser.pxi:1729: in lxml.etree._parseDoc (src/lxml/lxml.etree.c:116899)
    ???
src/lxml/parser.pxi:1063: in lxml.etree._BaseParser._parseUnicodeDoc (src/lxml/lxml.etree.c:110886)
    ???
src/lxml/parser.pxi:595: in lxml.etree._ParserContext._handleParseResultDoc (src/lxml/lxml.etree.c:105109)
    ???
src/lxml/parser.pxi:706: in lxml.etree._handleParseResult (src/lxml/lxml.etree.c:106817)
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   XMLSyntaxError: None (line 0)

src/lxml/parser.pxi:646: XMLSyntaxError
---------------------------------------- Captured stderr setup ----------------------------------------
INFO  [alembic.runtime.migration] Context impl SQLiteImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.runtime.migration] Running stamp_revision  -> 57f4c30b0ad4
INFO  [alembic.runtime.migration] Context impl SQLiteImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.runtime.migration] Running stamp_revision  -> df8d9f423f2b
INFO  [alembic.runtime.migration] Context impl SQLiteImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.runtime.migration] Running stamp_revision  -> 5a6db7a7222d
====================================== 2 failed in 2.92 seconds =======================================
```

Here are the results for the nexus

```
(.venv)naved001:~/mock-deploy-test/hil$ py.test tests/deployment/*networks.py
========================================= test session starts =========================================
platform linux2 -- Python 2.7.5, pytest-2.9.2, py-1.4.32, pluggy-0.3.1
rootdir: /home/naved001/mock-deploy-test/hil, inifile: setup.cfg
plugins: xdist-1.15.0, cov-1.8.1
collected 2 items 

tests/deployment/native_networks.py F
tests/deployment/vlan_networks.py F

============================================== FAILURES ===============================================
______________________________ TestNativeNetwork.test_isolated_networks _______________________________
self = <native_networks.TestNativeNetwork instance at 0x4235248>

    def test_isolated_networks(self):
    
        def create_networks():
            nodes = self.collect_nodes()
    
            # Create two networks
            network_create_simple('net-0', 'anvil-nextgen')
            network_create_simple('net-1', 'anvil-nextgen')
    
            ports = self.get_all_ports(nodes)
    
            # Assert that n0 and n1 are not on any network
            port_networks = self.get_port_networks(ports)
    
            assert self.get_network(nodes[0].nics[0].port, port_networks) == \
                set()
            assert self.get_network(nodes[1].nics[0].port, port_networks) == \
                set()
    
            # Connect n0 and n1 to net-0 and net-1 respectively
            api.node_connect_network(nodes[0].label,
                                     nodes[0].nics[0].label,
                                     'net-0')
            api.node_connect_network(nodes[1].label,
                                     nodes[1].nics[0].label,
                                     'net-1')
            deferred.apply_networking()
    
            # Assert that n0 and n1 are on isolated networks
            port_networks = self.get_port_networks(ports)
            assert self.get_network(nodes[0].nics[0].port, port_networks) == \
                set([nodes[0].nics[0].port])
            assert self.get_network(nodes[1].nics[0].port, port_networks) == \
                set([nodes[1].nics[0].port])
    
            # Add n2 and n3 to the same networks as n0 and n1 respectively
            api.node_connect_network(nodes[2].label,
                                     nodes[2].nics[0].label,
                                     'net-0')
            api.node_connect_network(nodes[3].label,
                                     nodes[3].nics[0].label,
                                     'net-1')
            deferred.apply_networking()
    
            # Assert that n2 and n3 have been added to n0 and n1's networks
            # respectively
            port_networks = self.get_port_networks(ports)
            assert self.get_network(nodes[0].nics[0].port, port_networks) == \
                set([nodes[0].nics[0].port, nodes[2].nics[0].port])
            assert self.get_network(nodes[1].nics[0].port, port_networks) == \
                set([nodes[1].nics[0].port, nodes[3].nics[0].port])
    
        def delete_networks():
            # Query the DB for nodes on this project
            project = api._must_find(model.Project, 'anvil-nextgen')
            nodes = project.nodes
            ports = self.get_all_ports(nodes)
    
            # Remove all nodes from their networks
            for node in nodes:
                attachment = model.NetworkAttachment.query \
                    .filter_by(nic=node.nics[0]).one()
                api.node_detach_network(node.label,
                                        node.nics[0].label,
                                        attachment.network.label)
            deferred.apply_networking()
    
            # Assert that none of the nodes are on any network
            port_networks = self.get_port_networks(ports)
            for node in nodes:
                assert self.get_network(node.nics[0].port, port_networks) == \
                    set()
    
            # Delete the networks
            api.network_delete('net-0')
            api.network_delete('net-1')
    
        # Create a project
        api.project_create('anvil-nextgen')
    
>       create_networks()

tests/deployment/native_networks.py:135: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
tests/deployment/native_networks.py:67: in create_networks
    port_networks = self.get_port_networks(ports)
haas/test_common.py:228: in get_port_networks
    switch_port_networks = session.get_port_networks(ports)
haas/ext/switches/nexus.py:171: in get_port_networks
    port_configs = self._port_configs(ports)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <haas.ext.switches.nexus._Session object at 0x47fe4d0>
ports = [<haas.model.Port object at 0x47fe110>, <haas.model.Port object at 0x47fcd10>, <haas.model.Port object at 0x460fcd0>, <haas.model.Port object at 0x47fe0d0>]

    def _port_configs(self, ports):
        alternatives = [
            re.escape(r'--More--'),
            r'Name:[^\n]*\n',
            r'  [A-Z][^:]*:[^\n]*\n',
            r'[\r\n]+.+# ',
        ]
        self.console.sendline('show int sw')
    
        # Find the first interface name
        self.console.expect(alternatives[1])
    
        _, interface = self.console.after.split(':', 1)
        interface = interface.strip()
        info = {interface: {}}
    
        while True:
            index = self.console.expect(alternatives)
            if index == 0:
                self.console.send(' ')
            elif index == 1:
                _, interface = self.console.after.split(':', 1)
                interface = interface.strip()
                info[interface] = {}
            elif index == 2:
                k, v = self.console.after.split(':', 1)
                info[interface][k.strip()] = v.strip()
            elif index == 3:
                break
    
        # The output of show int sw calls things "EthernetX/YY", but
        # everything else calls things "ethernet X/YY". Let's do the conversion
        # here.
        names_result = {}
        pattern = re.compile(r'Ethernet(\d+)/(\d+)')
        for k, v in info.iteritems():
            match = re.match(pattern, k)
            # TODO: We should verify there is in fact a match (there always
            # should be, but let's be paranoid).
>           switch, port = match.groups()
E           AttributeError: 'NoneType' object has no attribute 'groups'

haas/ext/switches/nexus.py:160: AttributeError
---------------------------------------- Captured stderr setup ----------------------------------------
INFO  [alembic.runtime.migration] Context impl SQLiteImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.runtime.migration] Running stamp_revision  -> 57f4c30b0ad4
INFO  [alembic.runtime.migration] Context impl SQLiteImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.runtime.migration] Running stamp_revision  -> df8d9f423f2b
_______________________________ TestNetworkVlan.test_isolated_networks ________________________________
self = <vlan_networks.TestNetworkVlan instance at 0x422f2d8>

    def test_isolated_networks(self):
    
        def get_legal_channels(network):
            response_body = api.show_network(network)
            response_body = json.loads(response_body)
            return response_body['channels']
    
        def create_networks():
            nodes = self.collect_nodes()
    
            # Create two networks
            network_create_simple('net-0', 'anvil-nextgen')
            network_create_simple('net-1', 'anvil-nextgen')
    
            ports = self.get_all_ports(nodes)
    
            # Assert that n0 and n1 are not on any network
            port_networks = self.get_port_networks(ports)
    
            assert self.get_network(nodes[0].nics[0].port, port_networks) == \
                set()
            assert self.get_network(nodes[1].nics[0].port, port_networks) == \
                set()
    
            # Get the channel ids for the tagged versions of the networks:
            net_tag = {}
            net_tag[0] = get_legal_channels('net-0')[1]
            net_tag[1] = get_legal_channels('net-1')[1]
    
            # Connect node 0 to net-0 (native mode)
            api.node_connect_network(nodes[0].label,
                                     nodes[0].nics[0].label,
                                     'net-0')
            # Connect node 1 to net-1 (tagged mode)
            api.node_connect_network(nodes[1].label,
                                     nodes[1].nics[0].label,
                                     'net-1',
                                     channel=net_tag[1])
            deferred.apply_networking()
    
            # Assert that n0 and n1 are on isolated networks
            port_networks = self.get_port_networks(ports)
            assert self.get_network(nodes[0].nics[0].port, port_networks) == \
                set([nodes[0].nics[0].port])
            assert self.get_network(nodes[1].nics[0].port, port_networks) == \
                set([nodes[1].nics[0].port])
    
            # Add n2 and n3 to the same networks as n0 and n1 respectively, but
            # with different channels (native vs. tagged)
            api.node_connect_network(nodes[2].label,
                                     nodes[2].nics[0].label,
                                     'net-0',
                                     channel=net_tag[0])
            api.node_connect_network(nodes[3].label,
                                     nodes[3].nics[0].label,
                                     'net-1')
            deferred.apply_networking()
    
            # Assert that n2 and n3 have been added to n0 and n1's networks
            # respectively
            port_networks = self.get_port_networks(ports)
            assert self.get_network(nodes[0].nics[0].port, port_networks) == \
                set([nodes[0].nics[0].port, nodes[2].nics[0].port])
            assert self.get_network(nodes[1].nics[0].port, port_networks) == \
                set([nodes[1].nics[0].port, nodes[3].nics[0].port])
    
            # Verify that we can put nodes on more than one network, with
            # different channels:
            api.node_connect_network(nodes[2].label,
                                     nodes[2].nics[0].label,
                                     'net-1')
            deferred.apply_networking()
            port_networks = self.get_port_networks(ports)
            assert self.get_network(nodes[1].nics[0].port, port_networks) == \
                set([nodes[1].nics[0].port,
                     nodes[2].nics[0].port,
                     nodes[3].nics[0].port])
    
        def delete_networks():
            # Query the DB for nodes on this project
            project = api._must_find(model.Project, 'anvil-nextgen')
            nodes = project.nodes
            ports = self.get_all_ports(nodes)
    
            # Remove all nodes from their networks. We first build up a list of
            # the arguments to the API calls, which has no direct references to
            # database objects, and then make the API calls and invoke
            # deferred.apply_networking after. This is important --
            # The API calls and apply_networking normally run in their own
            # transaction. We get away with not doing this in the tests because
            # we serialize everything ourselves, so there's no risk of
            # interference. If we were to hang on to references to database
            # objects across such calls however, things could get harry.
            all_attachments = []
            for node in nodes:
                attachments = model.NetworkAttachment.query \
                    .filter_by(nic=node.nics[0]).all()
                for attachment in attachments:
                    all_attachments.append((node.label,
                                            node.nics[0].label,
                                            attachment.network.label))
            for attachment in all_attachments:
                api.node_detach_network(*attachment)
                deferred.apply_networking()
    
            # Assert that none of the nodes are on any network
            port_networks = self.get_port_networks(ports)
            for node in nodes:
                assert self.get_network(node.nics[0].port, port_networks) == \
                    set()
    
            # Delete the networks
            api.network_delete('net-0')
            api.network_delete('net-1')
    
        # Create a project
        api.project_create('anvil-nextgen')
    
>       create_networks()

tests/deployment/vlan_networks.py:171: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
tests/deployment/vlan_networks.py:70: in create_networks
    port_networks = self.get_port_networks(ports)
haas/test_common.py:228: in get_port_networks
    switch_port_networks = session.get_port_networks(ports)
haas/ext/switches/nexus.py:171: in get_port_networks
    port_configs = self._port_configs(ports)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <haas.ext.switches.nexus._Session object at 0x4958990>
ports = [<haas.model.Port object at 0x4777990>, <haas.model.Port object at 0x477c510>, <haas.model.Port object at 0x477c890>, <haas.model.Port object at 0x4958750>]

    def _port_configs(self, ports):
        alternatives = [
            re.escape(r'--More--'),
            r'Name:[^\n]*\n',
            r'  [A-Z][^:]*:[^\n]*\n',
            r'[\r\n]+.+# ',
        ]
        self.console.sendline('show int sw')
    
        # Find the first interface name
        self.console.expect(alternatives[1])
    
        _, interface = self.console.after.split(':', 1)
        interface = interface.strip()
        info = {interface: {}}
    
        while True:
            index = self.console.expect(alternatives)
            if index == 0:
                self.console.send(' ')
            elif index == 1:
                _, interface = self.console.after.split(':', 1)
                interface = interface.strip()
                info[interface] = {}
            elif index == 2:
                k, v = self.console.after.split(':', 1)
                info[interface][k.strip()] = v.strip()
            elif index == 3:
                break
    
        # The output of show int sw calls things "EthernetX/YY", but
        # everything else calls things "ethernet X/YY". Let's do the conversion
        # here.
        names_result = {}
        pattern = re.compile(r'Ethernet(\d+)/(\d+)')
        for k, v in info.iteritems():
            match = re.match(pattern, k)
            # TODO: We should verify there is in fact a match (there always
            # should be, but let's be paranoid).
>           switch, port = match.groups()
E           AttributeError: 'NoneType' object has no attribute 'groups'

haas/ext/switches/nexus.py:160: AttributeError
---------------------------------------- Captured stderr setup ----------------------------------------
INFO  [alembic.runtime.migration] Context impl SQLiteImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.runtime.migration] Running stamp_revision  -> 57f4c30b0ad4
INFO  [alembic.runtime.migration] Context impl SQLiteImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.runtime.migration] Running stamp_revision  -> df8d9f423f2b
====================================== 2 failed in 10.59 seconds ======================================
```

--- a/haas/ext/switches/dell.py
+++ b/haas/ext/switches/dell.py
@@ -162,20 +162,17 @@ class _Session(_console.Session):
         """
 
         alternatives = [
-            re.escape(r'More: <space>, '
-                      'Quit: q or CTRL+Z, One line: <return> '),
+            r'More: .*',  # Prompt to press a key to continue
             r'Classification rules:\r\n',  # End
             r'[^ \t\r\n][^:]*:[^\n]*\n',   # Key:Value\r\n,
             r' [^\n]*\n',                  # continuation line (from k:v)
         ]
         self._sendline('show int sw %s' % interface)
-        # Find the first Key:Value pair (this is needed to skip past some
-        # possible matches for other patterns prior to this:
-        self.console.expect(alternatives[2])
 
+        # Name is the first field:
+        self.console.expect('Name: .*')
         k, v = self.console.after.split(':', 1)
         result = {k: v}
-
         while True:
             index = self.console.expect(alternatives)
             if index == 0:

--- a/haas/ext/switches/nexus.py
+++ b/haas/ext/switches/nexus.py
@@ -155,10 +155,13 @@ class _Session(_console.Session):
         pattern = re.compile(r'Ethernet(\d+)/(\d+)')
         for k, v in info.iteritems():
             match = re.match(pattern, k)
+            # sometimes switch has custom interface names, so we skip those
+            if match is None:
+                continue
             # TODO: We should verify there is in fact a match (there always
             # should be, but let's be paranoid).
             switch, port = match.groups()
-            names_result['ethernet %s/%s' % (switch, port)] = v
+            names_result['Ethernet%s/%s' % (switch, port)] = v
 
         result = {}
         for port in ports:

